### PR TITLE
SUBMARINE-499. Refactor job API to experiment API

### PR DESCRIPTION
**Pull request by jiwq**
### What is this PR for?
* Refactor the training job to experiment
* Redefined the experiment spec

### What type of PR is it?
Refactoring

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-499

### Questions:
* Does the licenses files need update? Yes/No
* Is there breaking changes for older versions? Yes/No
* Does this needs documentation? Yes/No

### CODE REVIEW

**Comment by jiwq** [comment]
Thanks @tangzhankun for the review. This is a draft PR which means it still work in progress, I'm still working for this.

**Comment by jiwq** [comment]
If i understand correctly, the environment's spec like below:
```json
"environment": {
    "name": "tensorflow",
    "dockerImage": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0",
    "cmd": [
         "python",
         "/var/tf_mnist/mnist_with_summaries.py",
         "--log_dir=/train/log",
         "--learning_rate=0.01",
         "--batch_size=150"
      ]
}
```
and the task level can overwrite the environment.

@wangdatan Sorry for the late delivery. The current pr failed by the Integration Testing caused by parse the response from submarine server. I will fix it ASAP and modify the propose.

**Comment by wangdatan** [comment]
No worries @jiwq . 

I think environment can be either uniquely named, or anonymously. See my comment on: 

https://issues.apache.org/jira/browse/SUBMARINE-507?focusedCommentId=17119816&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17119816 (you can search "One thing I realized is")

So I think we don't have to specify name and docker-image name at the same time. Also specify cmd in the environment doesn't make sense to me. Environment is supposed to be libraries and dependencies, not code. If user want to use code embedded inside an environment (docker image), that's their choice, but we may not want to add it to the environment API.

So the environment you listed I think it is sufficient to leave docker image for now: 

```
"environment": {
    "dockerImage": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
}
```

We can improve this part once SUBMARINE-507 finished.

**Comment by tangzhankun** [comment]
I spent some time checking the new Swagger Python API generated by this PR. Foud that it seems more complex than the previous JobSpec.

Here's the original JobSpec for tensorflow (Pytorch is very similar. The differences are taskSpec has {"master": ms", "worker":worker}):
```
jobLibrarySpec = JobLibrarySpec(name='tensorflow',
                                version='2.1.0',
                                image='gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0',
                                cmd='python /var/tf_mnist/mnist_with_summaries.py'
                                    ' --log_dir=/train/log --learning_rate=0.01'
                                    ' --batch_size=150',
                                env_vars={'ENV1': 'ENV1'})
ps = JobTaskSpec(name='ps',
                     image=None,
                     cmd=None,
                     env_vars=None,
                     resources='cpu=2,memory=1024M',
                     replicas=1)
worker = JobTaskSpec(name='worker',
                     image=None,
                     cmd=None,
                     env_vars=None,
                     resources='cpu=4,memory=2048M',
                     replicas=1)

body = JobSpec(name='mnist',
               namespace='submarine',
               library_spec=jobLibrarySpec,
               task_specs={"Ps": ps", "Worker": worker})
submarine_client.create_job(body=body)
```

Here's how we define an ad-hoc experiment spec for Tensorflow:

```
codeSpec = CodeSpec(mode = 'docker', 
                      url= 'gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0')
resourceSpec = ResourceSpec(cpu = '4', memory = '2048M')
psSpec = AdhocDistRoleSpec(replicas = 1,
                                name = 'ps',
                                cmd = 'Python ./train.py',
                                env_vars=None,
                                resource = resourceSpec
                                )
workerSpec = AdhocDistRoleSpec(replicas = 1,
                                name = 'ps',
                                cmd = 'Python ./train.py',
                                env_vars=None,
                                resource = resourceSpec
                                )
distTFSpec = AdhocDistTFSpec(image = None,
                                type = "tensorflow",
                                ps = psSpec,
                                worker = workerSpec,
                                chief = None,
                                evaluator = None,
                                )
body = AdhocExperiemntSpec(name = 'minist',
                                namespace = 'default',
                                code = codeSpec,
                                type = 'adhoc',
                                spec = distTFSpec
                                )
submarine_client.create_experiment(body=body)
```

Here's how we define an ad-hoc experiment spec for PyTorch:
```
codeSpec = CodeSpec(mode = 'docker', 
                      url= 'gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0')
resourceSpec = ResourceSpec(cpu = '4', memory = '2048M')
masterSpec = AdhocScriptSpec(replicas = 1,
                                name = 'ps',
                                cmd = 'Python ./train.py',
                                env_vars=None,
                                resource = resourceSpec,
                                image = Node)
workerSpec = AdhocScriptSpec(replicas = 1,
                                name = 'ps',
                                cmd = 'Python ./train.py',
                                env_vars=None,
                                resource = resourceSpec,
                                image = Node)
distPYSpec = AdhocDistPyTorchSpec(image = None,
                                type = "pytorch",
                                master = masterSpec,
                                worker = workerSpec)
body = AdhocExperiemntSpec(name = 'minist',
                                namespace = 'default',
                                code = codeSpec,
                                type = 'adhoc',
                                spec = distTFSpec)
submarine_client.create_experiment(body=body)
```
@jiwq , @wangdatan , @kevin85421  Please check it to have a feeling of the generated API.

I'm fine if we go with this PR, but we should use `AdhocDistRoleSpec` for PyTorch as well as Tensorflow instead of `AdhocScriptSpec`. It's too confusing.

**Comment by tangzhankun** [comment]
And this is JSON format of this PR:

```
{
  "name": "mnist",
  "namespace": "default",
  "type": "adhoc",
  "spec": {
    "type": "tensorflow",
    "ps": {
      "name": "tensorflow",
      "cmd": [
        "python",
        "/var/tf_mnist/mnist_with_summaries.py",
        "--log_dir=/train/log",
        "--learning_rate=0.01",
        "--batch_size=150"
      ],
      "envVars": {
        "ENV_1": "ENV1"
      },
      "resource": {
        "cpu": "1",
        "memory": "512M"
      },
      "replicas": 1
    },
    "worker": {
      "name": "tensorflow",
      "cmd": [
        "python",
        "/var/tf_mnist/mnist_with_summaries.py",
        "--log_dir=/train/log",
        "--learning_rate=0.01",
        "--batch_size=150"
      ],
      "envVars": {
        "ENV_1": "ENV1"
      },
      "resource": {
        "cpu": "1",
        "memory": "512M"
      },
      "replicas": 1
    }
  },
  "code": {
    "mode": "docker",
    "url": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
  }
}
```

And our Job Spec in JSON format:
```
{
  "name": "mnist",
  "librarySpec": {
    "name": "TensorFlow",
    "version": "2.1.0",
    "image": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0",
    "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150",
    "envVars": {
      "ENV_1": "ENV1"
    }
  },
  "submitterSpec": {
    "type": "k8s",
    "namespace": "submarine"
  },
  "taskSpecs": {
    "Ps": {
      "name": "tensorflow",
      "replicas": 2,
      "resources": "cpu=4,memory=2048M,nvidia.com/gpu=1"
    },
    "Worker": {
      "name": "tensorflow",
      "replicas": 2,
      "resources": "cpu=4,memory=2048M,nvidia.com/gpu=1"
    }
  }
}
```




And our new expected experiment spec:
```
{
  "name": "mnist",
  "namespace": "default",
  “type”: “adhoc”,
  "environment": {
      "dockerImage": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
  }
  "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150",
  "spec": {
    “type”: "TensorFlow",
    "Ps": {
      "name": "tensorflow",
      "replicas": 2,
      "resources": "cpu=4,memory=2048M,nvidia.com/gpu=1"
    },
    "Worker": {
      "name": "tensorflow",
      "replicas": 2,
      "resources": "cpu=4,memory=2048M,nvidia.com/gpu=1"
    }
  }
```

**Comment by wangdatan** [comment]
Thanks @tangzhankun , the last one 

> And my understanding of the new expected experiment spec in the discussion. @jiwq , 
> @wangdatan , @kevin85421 do you feel the below is better?

Is way better to me, even if it may not be exactly same as the final API, it looks very similar to what we want, so we can release it in the 0.4.0.

**Comment by jiwq** [comment]
Thanks @tangzhankun , the last one is better to me. If no more comments, I'll do it in 0.4.0 and release it soon.

**Comment by jiwq** [comment]
When I try to refactor this structure, I found the spec can't deserialize to Java Map. So I think we can do some modify. The experiment spec like:
```json
{
  "meta": {
    "name": "mnist",
    "namespace": "submarine",
    "framework": "TensorFlow",
    "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150"
  },
  "environment": {
    "dockerImage": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
  },
  "spec": {
    "Ps": {
      "replicas": 1,
      "resources": "cpu=4,memory=2048M"
    },
    "Worker": {
      "replicas": 2,
      "resources": "cpu=2,memory=1024M,nvidia.com/gpu=1"
    }
  }
}
```
Python client:
```python
meta = Meta(name='mnist', namespace='submarine', framework='TensorFlow', 
           cmd: "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150")
environment = Environment(docker-image='gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0')
ps = Task(environment = None, cmd=None, env_vars=None,
           resources='cpu=2,memory=1024M', replicas=1)
worker = TaskSpec(environment = None, cmd=None,  env_vars=None,
           resources='cpu=4,memory=2048M', replicas=2)
body = ExperimentSpec(meta=meta, environment = environment,
           spec={"type":"tensorflow", "Ps": ps, "Worker": worker})
client.create_experiment(body=body)
```

**Comment by tangzhankun** [comment]
@jiwq , thanks for the update. I think this spec looks good to me too.

```
{
  "meta": {
    "name": "mnist",
    "namespace": "submarine",
    "framework": "TensorFlow",
    "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150"
  },
  "environment": {
    "dockerImage": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
  },
  "spec": {
    "Ps": {
      "replicas": 1,
      "resources": "cpu=4,memory=2048M"
    },
    "Worker": {
      "replicas": 2,
      "resources": "cpu=2,memory=1024M,nvidia.com/gpu=1"
    }
  }
}
```

**Comment by pingsutw** [comment]
> My biggest worry is, does it mean we need to double our efforts since we will do two big changes (rename Job to Experiment, and move old Experiment spec to new) to our code and SDK? If it is not two big changes, I will be fine with this proposal.

I have written a script to auto-generate client SDK code, so there is less effort to refactor.

```
{
  "meta": {
    "name": "mnist",
    "namespace": "submarine",
    "framework": "TensorFlow",
    "cmd": "python /var/tf_mnist/mnist_with_summaries.py --log_dir=/train/log --learning_rate=0.01 --batch_size=150"
  },
  "environment": {
    "dockerImage": "gcr.io/kubeflow-ci/tf-mnist-with-summaries:1.0"
  },
  "spec": {
    "Ps": {
      "replicas": 1,
      "resources": "cpu=4,memory=2048M"
    },
    "Worker": {
      "replicas": 2,
      "resources": "cpu=2,memory=1024M,nvidia.com/gpu=1"
    }
  }
}
```
The new job spec is very clear to me.
After this PR is merged, I will update SDK ASAP.

**Comment by wangdatan** [comment]
Thanks @jiwq! 

The latest spec looks good to me, even it may not be exactly same as what we will release in 0.5 or after, but it is shaped in a similar way which we can migrate them to a more stable version in 0.5. 

Just one thing, I think we need to highlight now Experiment API is in alpha stage which is subjected to incompatible changes in the future releases. We need to target to move it to beta version from 0.5 release. 

I will rely on other folks to do detailed code reviews.

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/rest/ExperimentRestApi.java [hunk 1]
```diff
@@ -70,147 +70,150 @@ public Response ping() {
...
-      return parseJobServiceException(e);
+      return parseExperimentServiceException(e);
     }
   }
 
   @GET
   @Path("/logs")
-  @Operation(summary = "Log jobs",
-          tags = {"jobs"},
+  @Operation(summary = "List experiment's all log",
```
Comment (tangzhankun) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/rest/ExperimentRestApi.java:181]:
```suggestion
  @Operation(summary = "List experiment's log",
```

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/K8sSubmitter.java [hunk 2]
```diff
@@ -186,59 +181,59 @@ private Job parseResponseObject(Object object, ParseOp op) throws SubmarineRunti
...
+          getJobLabelSelector(spec), null, null,
           null, null);
       for (V1Pod pod: podList.getItems()) {
         String podName = pod.getMetadata().getName();
-        jobLog.addPodLog(podName, null);
+        experimentLog.addPodLog(podName, null);
       }
     } catch (final ApiException e) {
-      LOG.error("Error when listing pod for job:" + jobSpec.getName(), e.getMessage());
+      LOG.error("Error when listing pod for job:" + spec.getMeta().getName(), e.getMessage());
```
Comment (tangzhankun) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/K8sSubmitter.java:198]:
```suggestion
      LOG.error("Error when listing pod for experiment:" + spec.getMeta().getName(), e.getMessage());
```

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/K8sSubmitter.java [hunk 3]
```diff
@@ -186,59 +181,59 @@ private Job parseResponseObject(Object object, ParseOp op) throws SubmarineRunti
...
-            Integer.MAX_VALUE, null, Boolean.FALSE, 
+            Integer.MAX_VALUE, null, Boolean.FALSE,
             Integer.MAX_VALUE, null, Boolean.FALSE);
 
-        jobLog.addPodLog(podName, podLog);
+        experimentLog.addPodLog(podName, podLog);
       }
     } catch (final ApiException e) {
-      LOG.error("Error when listing pod for job:" + jobSpec.getName(), e.getMessage());
+      LOG.error("Error when listing pod for job:" + spec.getMeta().getName(), e.getMessage());
```
Comment (tangzhankun) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/K8sSubmitter.java:225]:
```suggestion
      LOG.error("Error when listing pod for experiment:" + spec.getMeta().getName(), e.getMessage());
```

File: submarine-test/test-k8s/src/test/java/org/apache/submarine/rest/ExperimentRestApiIT.java [hunk 4]
```diff
@@ -161,66 +161,66 @@ private void run(String body, String patchBody, String contentType) throws Excep
...
+    KfOperator operator = kfOperatorMap.get(experiment.getSpec().getMeta().getFramework()
         .toLowerCase());
     JsonObject rootObject = getJobByK8sApi(operator.getGroup(), operator.getVersion(),
-        operator.getNamespace(), operator.getPlural(), job.getName());
+        operator.getNamespace(), operator.getPlural(), experiment.getName());
     JsonObject metadataObject = rootObject.getAsJsonObject("metadata");
 
     String uid = metadataObject.getAsJsonPrimitive("uid").getAsString();
-    LOG.info("Uid from Job REST is {}", job.getUid());
+    LOG.info("Uid from Job REST is {}", experiment.getUid());
```
Comment (tangzhankun) [inline submarine-test/test-k8s/src/test/java/org/apache/submarine/rest/ExperimentRestApiIT.java:200]:
```suggestion
    LOG.info("Uid from Experiment REST is {}", experiment.getUid());
```

File: submarine-test/test-k8s/src/test/java/org/apache/submarine/rest/ExperimentRestApiIT.java [hunk 5]
```diff
@@ -132,26 +132,26 @@ public void testPyTorchWithYamlSpec() throws Exception {
   private void run(String body, String patchBody, String contentType) throws Exception {
     // create
     LOG.info("Create training job by Job REST API");
```
Comment (tangzhankun) [inline submarine-test/test-k8s/src/test/java/org/apache/submarine/rest/ExperimentRestApiIT.java:134]:
```suggestion
    LOG.info("Create training job by Experiment REST API");
```

File: submarine-test/test-k8s/src/test/java/org/apache/submarine/rest/ExperimentRestApiIT.java [hunk 6]
```diff
@@ -161,66 +161,66 @@ private void run(String body, String patchBody, String contentType) throws Excep
...
     LOG.info("Uid from K8s REST is {}", uid);
-    Assert.assertEquals(job.getUid(), uid);
+    Assert.assertEquals(experiment.getUid(), uid);
 
     String creationTimestamp = metadataObject.getAsJsonPrimitive("creationTimestamp")
         .getAsString();
-    Date expectedDate = new DateTime(job.getAcceptedTime()).toDate();
+    Date expectedDate = new DateTime(experiment.getAcceptedTime()).toDate();
     Date actualDate = new DateTime(creationTimestamp).toDate();
     LOG.info("CreationTimestamp from Job REST is {}", expectedDate);
```
Comment (tangzhankun) [inline submarine-test/test-k8s/src/test/java/org/apache/submarine/rest/ExperimentRestApiIT.java:208]:
```suggestion
    LOG.info("CreationTimestamp from Experiment REST is {}", expectedDate);
```

File: docs/userdocs/k8s/README.md [hunk 7]
```diff
@@ -18,24 +18,21 @@ under the License.
 -->
 
 # Running Submarine on K8s
-
 Submarine for K8s supports standalone distributed TensorFlow and PyTorch.
```
Comment (tangzhankun) [inline docs/userdocs/k8s/README.md:21]:
```suggestion
Submarine for K8s supports distributed TensorFlow and PyTorch.
```

File: docs/userdocs/k8s/README.md [hunk 8]
```diff
@@ -18,24 +18,21 @@ under the License.
 -->
 
 # Running Submarine on K8s
```
Comment (tangzhankun) [inline docs/userdocs/k8s/README.md:20]:
```suggestion
# Submarine on K8s
```

File: docs/userdocs/k8s/README.md [hunk 9]
```diff
@@ -18,24 +18,21 @@ under the License.
...
-
-## Submarine on K8s guide
-
-### Prepare K8s and deploy Submarine Service
-
-[Setup Kubernetes](setup-kubernetes.md): Submarine can be deployed on any K8s environment if version matches. If you don't have a running K8s, you can follow the steps to set up a K8s using [kind, Kubernetes-in-Docker](https://kind.sigs.k8s.io/) for testing purpose.
+Submarine can run on K8s >= 1.14, supports features like GPU isolation.
 
-After you have an up-and-running K8s, you can follow [Deploy Submarine Services on K8s](deploy-submarine.md) guide to deploy Submarine services on K8s using Helmchart in minutes (FIXME: is it true?).
+## Installing Submarine
```
Comment (tangzhankun) [inline docs/userdocs/k8s/README.md:25]:
```suggestion
## Install Submarine
```

File: docs/userdocs/k8s/README.md [hunk 10]
```diff
@@ -18,24 +18,21 @@ under the License.
...
+Submarine can be deployed on any K8s environment if version matches. If you don't have a running K8s, you can follow the steps to set up a K8s using [kind, Kubernetes-in-Docker](https://kind.sigs.k8s.io/) for testing purpose, we provides simple [tutorial](kind.md).
 
-### Use Submarine
+### Using Helm Charts
```
Comment (tangzhankun) [inline docs/userdocs/k8s/README.md:28]:
```suggestion
### Use Helm Charts
```

File: docs/userdocs/k8s/README.md [hunk 11]
```diff
@@ -18,24 +18,21 @@ under the License.
...
+## Use Submarine
 
-- [Run model training using Tensorflow](run-tensorflow-on-k8s.md)
-- [Run model training using PyTorch](FIXME, add one).
+### Model training (experiment) on K8s
+- [Run model training using Tensorflow](run-tensorflow-experiment.md)
+- [Run model training using PyTorch](run-pytorch-experiment.md)
+- [Experiment API Reference](api/experiment.md)
 
 ## References
```
Comment (tangzhankun) [inline docs/userdocs/k8s/README.md:38]:
```suggestion

```

File: docs/userdocs/k8s/api/experiment.md [hunk 12]
```diff
@@ -17,216 +17,152 @@ specific language governing permissions and limitations
 under the License.
 -->
 
-# Submarine Server Guide
-This guide covers the deploy and running the training job by submarine server.
-It now supports Tensorflow and PyTorch jobs.
+# Experiment API Reference
```
Comment (tangzhankun) [inline docs/userdocs/k8s/api/experiment.md:21]: We should add below note:
> Note:
> The Experiment API is in the alpha stage which is subjected to incompatible changes in future releases.

**Comment by tangzhankun** [comment]
@jiwq Thanks for the update! I'll commit this soon.

**Pull request closed without merging**